### PR TITLE
Ability to use containerd as runtime in kube-up (on Ubuntu)

### PR DESCRIPTION
**Pull request by dims**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
We need a job that uses containerd, both in master and node. This PR enables us to install containerd automatically when we use a ubuntu OS base image if it is not present on the image. This also enables us to inject newer versions of containerd and runc to help us test against them as well. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -2686,6 +2681,16 @@ oom_score = -999
 [plugins.cri.registry.mirrors."docker.io"]
   endpoint = ["https://mirror.gcr.io","https://registry-1.docker.io"]
 EOF
+
+  # Reuse docker group for containerd.
+  local containerd_gid="$(cat /etc/group | grep ^docker: | cut -d: -f 3)"
```
Comment (dims) [inline cluster/gce/gci/configure-helper.sh]: before this change the script was assuming that `docker` is always installed!

File: cluster/gce/gci/configure.sh [hunk 2]
```diff
@@ -436,6 +443,56 @@ function install-docker {
...
+    "deb [arch=amd64] https://download.docker.com/linux/$(. /etc/os-release; echo "$ID") \
+    $(lsb_release -cs) stable"
+
+  # Install containerd from Docker repo
+  apt-get update && \
+    apt-get install -y --no-install-recommends containerd
+  rm -rf /var/lib/apt/lists/*
+
+
+  # Override to latest versions of containerd and runc
```
Comment (dims) [inline cluster/gce/gci/configure.sh]: the artifacts we download for containerd does have say the systemd service definition, so we basically install a version of containerd and then swap out the binaries

File: cluster/gce/gci/configure.sh [hunk 3]
```diff
@@ -33,6 +33,13 @@ DEFAULT_CRICTL_SHA1="5c18f4e52ab524d429063b78d086dd18b894aae7"
 DEFAULT_MOUNTER_TAR_SHA="8003b798cf33c7f91320cd6ee5cec4fa22244571"
 ###
 
+# Ability to inject custom versions (Ubuntu OS images ONLY)
+# if KUBE_CONTAINERD_VERSION_OVERRIDE or KUBE_RUNC_VERSION_OVERRIDE
+# is set to empty then we do not override the version(s) and just
+# use whatever is in the default installation of containerd package
+CONTAINERD_VERSION=${KUBE_CONTAINERD_VERSION_OVERRIDE="v1.3.3"}
+RUNC_VERSION=${KUBE_RUNC_VERSION_OVERRIDE="v1.0.0-rc10"}
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: is this the right syntax? I'm used to
```bash
CONTAINERD_VERSION=${KUBE_CONTAINERD_VERSION_OVERRIDE:-"v1.3.3"}
RUNC_VERSION=${KUBE_RUNC_VERSION_OVERRIDE:-"v1.0.0-rc10"}
```

File: cluster/gce/gci/configure.sh [hunk 4]
```diff
@@ -33,6 +33,13 @@ DEFAULT_CRICTL_SHA1="5c18f4e52ab524d429063b78d086dd18b894aae7"
 DEFAULT_MOUNTER_TAR_SHA="8003b798cf33c7f91320cd6ee5cec4fa22244571"
 ###
 
+# Ability to inject custom versions (Ubuntu OS images ONLY)
+# if KUBE_CONTAINERD_VERSION_OVERRIDE or KUBE_RUNC_VERSION_OVERRIDE
+# is set to empty then we do not override the version(s) and just
+# use whatever is in the default installation of containerd package
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: I would have expected this to be flipped... by default, use the installed versions, and require overrides to actively pull/install packages during CI

Comment (liggitt) [reply cluster/gce/gci/configure.sh]: also, if this install process is ubuntu only, and triggers install, maybe name the envvars to match that, e.g. `UBUNTU_INSTALL_CONTAINERD_VERSION` and `UBUNTU_INSTALL_RUNC_VERSION`?

File: cluster/gce/gci/configure.sh [hunk 5]
```diff
@@ -448,11 +505,17 @@ function ensure-container-runtime {
     fi
     docker version
   elif [[ "${container_runtime}" == "containerd" ]]; then
+    set -x
     if ! command -v ctr >/dev/null 2>&1; then
-      echo "ERROR ctr not found. Aborting."
-      exit 2
+      install-containerd
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: name this so it is clear that it may be a no-op, and is ubuntu-only?

File: cluster/gce/gci/configure.sh [hunk 6]
```diff
@@ -436,6 +443,56 @@ function install-docker {
   rm -rf /var/lib/apt/lists/*
 }
 
+# If we are on ubuntu we can try to install containerd
+function install-containerd {
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: return early if UBUNTU_INSTALL_CONTAINERD_VERSION and UBUNTU_INSTALL_RUNC_VERSION are unset so we don't do package install work unnecessarily?

File: cluster/gce/gci/configure.sh [hunk 7]
```diff
@@ -436,6 +443,56 @@ function install-docker {
...
+  # bailout if we are not on ubuntu
+  if ! command -v apt-get >/dev/null 2>&1; then
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: if we get here, it means UBUNTU_INSTALL_CONTAINERD_VERSION or UBUNTU_INSTALL_RUNC_VERSION are set... should we error in non-ubuntu/non-amd64 envs? it means someone is running this requesting those be installed, and they aren't

File: cluster/gce/gci/configure.sh (hunk 5 above)
Comment (liggitt) [reply cluster/gce/gci/configure.sh]: why do we only call install-containerd if `ctr` can't be found?

File: cluster/gce/gci/configure.sh [hunk 8]
```diff
@@ -448,11 +505,17 @@ function ensure-container-runtime {
     fi
     docker version
   elif [[ "${container_runtime}" == "containerd" ]]; then
+    set -x
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: why is this needed?

File: cluster/gce/gci/configure.sh (hunk 3 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: yep. we use the same technique in [`GOLDFLAGS`](https://github.com/kubernetes-sigs/cluster-capacity/blob/master/vendor/k8s.io/kubernetes/hack/lib/golang.sh#L780-L784) basically allows someone to clear the flags by using `GOLDFLAGS=` (environment variable is present but is set to empty scenario)

File: cluster/gce/gci/configure.sh (hunk 4 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: i wanted the new CI job to test the latest from containerd (1.3.3) and runc (v10) by default, these are not available currently. 

yep, we can change it around once i get the CI jobs go green (and tweak what we want be defaults and env vars)

File: cluster/gce/gci/configure.sh (hunk 5 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: if the image already has `ctr` the assumption is that the user is specifying an image that already has the things they want to test (like `COS`), so we don't try to do more in this case. (same logic we had in the docker install)

File: cluster/gce/gci/configure.sh (hunk 6 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: yep, can do that when i get this version working

File: cluster/gce/gci/configure.sh (hunk 8 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: yep will remove the `set -x`

File: cluster/gce/gci/configure.sh (hunk 7 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: i can add a explicit check for ubuntu

File: cluster/gce/gci/configure.sh (hunk 5 above)
Comment (liggitt) [reply cluster/gce/gci/configure.sh]: ah, I missed this is matching an existing pattern

Comment (dims) [reply cluster/gce/gci/configure.sh]: let me change the checks from `ctr` to `containerd`

Comment (liggitt) [reply cluster/gce/gci/configure.sh]: that makes the envvars misleading... if you happen to use an image that already has the binary installed, setting the envvars won't actually cause that version to be used

File: cluster/gce/gci/configure.sh (hunk 6 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: Done

File: cluster/gce/gci/configure.sh (hunk 5 above)
Comment (liggitt) [reply cluster/gce/gci/configure.sh]: would an approach like this be easier to understand?

```bash
elif [[ "${container_runtime}" == "containerd" ]]; then
  # Install containerd/runc if requested
  if [[ ! -z "${INSTALL_CONTAINERD_VERSION:-}" || ! -z "${INSTALL_RUNC_VERSION}" ]]; then
    install-containerd
  fi
  
  # Verify presence and print versions of ctr, containerd, runc
  if ! command -v ctr >/dev/null 2>&1; then
    echo "ERROR ctr not found. Aborting."
  fi
  ctr --version
  
  if ! command -v containerd >/dev/null 2>&1; then
    echo "ERROR containerd not found. Aborting."
  fi
  containerd --version
  
  if ! command -v runc >/dev/null 2>&1; then
    echo "ERROR runc not found. Aborting."
  fi
  runc --version
fi
```

File: cluster/gce/gci/configure.sh (hunk 4 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: making the change now. just going to use `UBUNTU_INSTALL_CONTAINERD_VERSION` and `UBUNTU_INSTALL_RUNC_VERSION` everywhere

File: cluster/gce/gci/configure.sh (hunk 5 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: fixing!

File: cluster/gce/gci/configure.sh [hunk 9]
```diff
@@ -33,6 +33,13 @@ DEFAULT_CRICTL_SHA1="5c18f4e52ab524d429063b78d086dd18b894aae7"
 DEFAULT_MOUNTER_TAR_SHA="8003b798cf33c7f91320cd6ee5cec4fa22244571"
 ###
 
+# Ability to inject custom versions (Ubuntu OS images ONLY)
+# if KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION or KUBE_UBUNTU_INSTALL_RUNC_VERSION
+# is set to empty then we do not override the version(s) and just
+# use whatever is in the default installation of containerd package
+UBUNTU_INSTALL_CONTAINERD_VERSION=${KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION="v1.3.3"}
+UBUNTU_INSTALL_RUNC_VERSION=${KUBE_UBUNTU_INSTALL_RUNC_VERSION="v1.0.0-rc10"}
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: comment about defaulting to use already-installed packages is still outstanding.

File: cluster/gce/gci/configure.sh [hunk 10]
```diff
@@ -436,6 +443,56 @@ function install-docker {
...
+
+  # Install containerd from Docker repo
+  apt-get update && \
+    apt-get install -y --no-install-recommends containerd
+  rm -rf /var/lib/apt/lists/*
+
+
+  # Override to latest versions of containerd and runc
+  systemctl stop containerd
+  if [[ ! -z "${CONTAINERD_VERSION:-}" ]]; then
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: UBUNTU_INSTALL_CONTAINERD_VERSION everywhere

File: cluster/gce/gci/configure.sh [hunk 11]
```diff
@@ -436,6 +443,56 @@ function install-docker {
...
+    curl -fsSL "https://github.com/containerd/containerd/releases/download/${CONTAINERD_VERSION}/containerd-${CONTAINERD_VERSION:1}.linux-amd64.tar.gz" | tar --overwrite -xzv -C /usr/
+  fi
+  if [[ ! -z "${RUNC_VERSION:-}" ]]; then
```
Comment (liggitt) [inline cluster/gce/gci/configure.sh]: UBUNTU_INSTALL_RUNC_VERSION everywhere

File: cluster/gce/gci/configure.sh (hunk 9 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: Done!

File: cluster/gce/gci/configure.sh (hunk 10 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: Done

File: cluster/gce/gci/configure.sh (hunk 11 above)
Comment (dims) [reply cluster/gce/gci/configure.sh]: Done

**Comment by liggitt** [comment]
thanks, this seems a lot more straightforward

/hold cancel

**Comment by liggitt** [comment]
looks good to me, will defer official lgtm to someone more bash-conversant

File: cluster/gce/gci/configure-helper.sh (hunk 1 above)
Comment (BenTheElder) [reply cluster/gce/gci/configure-helper.sh]: IIRC this is in fact true on the images we use?

File: cluster/gce/gci/configure-helper.sh [hunk 12]
```diff
@@ -2686,6 +2681,16 @@ oom_score = -999
...
+  if [[ ! -z "${containerd_gid:-}" ]]; then
+    cat >> "${config_path}" <<EOF
+# reuse id of the docker group
+[grpc]
+  gid = ${containerd_gid}
+EOF
+  fi
```
Comment (liggitt) [inline cluster/gce/gci/configure-helper.sh]: do we need an else? if docker is not around, what gid is used?

**Comment by dims** [comment]
/test pull-kubernetes-e2e-gce

2 flakes!

File: cluster/gce/gci/configure-helper.sh (hunk 12 above)
Comment (dims) [reply cluster/gce/gci/configure-helper.sh]: @liggitt we let containerd default to whatever it needs to. 

https://github.com/containerd/containerd/blame/master/docs/man/containerd-config.toml.5.md#L45-L49

**Comment by aojea** [comment]
`containerd` depends on a specific `runc` version for advanced features, actually they are vendoring it
https://github.com/containerd/containerd/blob/master/RUNC.md
maybe is not a good idea to have the possibility to use different versions of `runc` and `containerd`, and we should pick always the one provided by the `containerd` version

This is how it is done in KIND
https://github.com/kubernetes-sigs/kind/blob/master/images/base/Dockerfile#L80-L90

**Comment by dims** [comment]
@aojea you can certainly specify just `KUBE_UBUNTU_INSTALL_CONTAINERD_VERSION` and that will reuse the `runc` installed by containerd. So no issues.

**Comment by dims** [comment]
the RUNC switch is needed for scenarios we encountered recently namely - https://github.com/kubernetes/kubernetes/issues/86312

**Comment by dims** [comment]
/test pull-kubernetes-e2e-gce

3 flakes!

**Comment by dims** [comment]
/test pull-kubernetes-e2e-gce

1 flake!

**Comment by dims** [comment]
@justaugustus @cblecker @BenTheElder 

this is ready to merge, tested using another [WIP PR](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/87654/pull-kubernetes-e2e-gce-ubuntu/1226333916411990016/artifacts/e2e-f953cae636-3cd98-master/kube-master-installation.log):
```
Feb 09 02:51:43.031489 e2e-f953cae636-3cd98-master configure.sh[2096]: ctr github.com/containerd/containerd v1.3.3
Feb 09 02:51:43.111827 e2e-f953cae636-3cd98-master configure.sh[2096]: containerd github.com/containerd/containerd v1.3.3 d76c121f76a5fc8a462dc64594aea72fe18e1178
Feb 09 02:51:43.121796 e2e-f953cae636-3cd98-master configure.sh[2096]: runc version 1.0.0-rc10
Feb 09 02:51:43.121796 e2e-f953cae636-3cd98-master configure.sh[2096]: spec: 1.0.1-dev
```

**Comment by dims** [comment]
/test pull-kubernetes-node-e2e

flake:
`[k8s.io] Summary API [NodeConformance] when querying /stats/summary should report resource usage through the stats api`
